epc_data_analysis/pipeline/feature_engineering.py

```diff
@@ -0,0 +1,227 @@
+    df = df[df.DIFF_POT_ENERGY_RATING >= 0.0]
+
+    return df
+
+
+def map_quality_to_number(df, list_of_features):
+
+    quality_to_num_dict = {
+        "Very Good": 5.0,
+        "Good": 4.0,
+        "Average": 3.0,
+        "Poor": 2.0,
+        "Very Poor": 1.0,
+    }
+
+    for feature in list_of_features:
+        df[feature + "_AS_NUM"] = df[feature].map(quality_to_num_dict)
+
+    return df
+
+
+def get_heating_features(df, fine_grained_HP_types=False):
+    """Get heating type category based on HEATING_TYPE category.
+    heating_system: heat pump, boiler, community scheme etc.
+    heating_source: oil, gas, LPC, electric.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        Dataframe that is updated with heating features.
+
+    fine_grained_HP_types : bool, default=False
+        If True, get different heat pump types (air sourced, ground sourced etc.).
+        If False, return "heat pump" as heating type category.
+
+    Return
+    ---------
+    df : pandas.DataFrame
+        Updated dataframe with heating system and source."""
+
+    # Collections
+    heating_system_types = []
+    heating_source_types = []
+
+    # Get heating types
+    heating_types = df["MAINHEAT_DESCRIPTION"]
+
+    # Get specific and general heating category for each entry
+    for heating in heating_types:
+
+        # Set default value
+        system_type = "unknown"
+        source_type = "unknown"
+
+        # If heating value exists
+        if not pd.isnull(heating):
+
+            # Lowercase
+            heating = heating.lower()
+
+            # Different heat pump types
+            # --------------------------
+
+            if "ground source heat pump" in heating:
+                system_type = "ground source heat pump"
+                source_type = "electric"
+
+            elif "air source heat pump" in heating:
+                system_type = "air source heat pump"
+                source_type = "electric"
+
+            elif "water source heat pump" in heating:
+                system_type = "water source heat pump"
+                source_type = "electric"
+
+            elif "heat pump" in heating:
+                system_type = "heat pump"
+                source_type = "electric"
+
+            # Electric heaters
+            # --------------------------
+
+            elif "electric storage heaters" in heating:
+                system_type = "storage heater"
+                source_type = "electric"
+
+            elif "electric underfloor heating" in heating:
+                system_type = "underfloor heating"
+                source_type = "electric"
+
+            # Warm air
+            # --------------------------
+
+            elif "warm air" in heating:
+                system_type = "warm air"
+                source_type = "electric"
+
+            # Boiler and radiator / Boiler and underfloor / Community scheme / Heater (unspecified)
+            # --------------------------
+
+            elif (
```

Comment:
Could use `any` instead of long conditional statements

Reply:
Worked like a charm with `any` and is much more readable now. Thanks for this tip!